### ISSUE TITLE
[Documentation] Explain the implications of typedef - "parameter" and "arrow-parameter"

### ISSUE THREAD

**Issue by aluanhaddad** [palantir/tslint#3502]
I believe it would be beneficial to document the hazards of these options as they are likely to trip up newcomers.

For example, using a configuration such as
```json
{
  "extends": "tslint:recommended",
  "rules": {
    "typedef": [true, "parameter", "arrow-parameter"]
  }
}
```
introduces serious type safety hazards in the context of callbacks.

Consider a library containing a declaration such as
```ts
export declare function processData(callback: (data: any) => void): void;
```
and consuming code such as
```ts
import {processData} from 'lib';

processData(data => {
  console.log(data.method());
});
```
with the above configuration, TS Lint will raise
```sql
expected arrow-parameter: 'data' to have a typedef (typedef)
```
This may prompt a rewrite, especially if the fantastic (:heart:) `no-any` rule is also enabled, to something like
```ts
processData((data: Response) => {
  console.log(data.json());
});
```
This creates a false sense of security for the author and future maintainers.

While annotating the types of callback parameters is occasionally necessary, such as when working with advanced function composition facilities provided by libraries such as Ramda, I think that far and away it should be the exception rather than the rule.

My opinion notwithstanding, the trade-offs should be documented.

I would be happy to submit a PR if you are interested.

Thanks for TS Lint!

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!
